django/curator/tag_deduplication.py

```diff
@@ -0,0 +1,162 @@
+            labelled_examples["distinct"] = pairs
+        else:
+            labelled_examples["match"] = pairs
+
+        self.deduper.mark_pairs(labelled_examples)
+
+    def prepare_training_data(self):
+        tags = Tag.objects.all().values()
+        data = {row["id"]: row for i, row in enumerate(tags)}
+        return data
+
+    def console_label(self):
+        dedupe.console_label(self.deduper)
+
+
+# This class is used to help to make the initial canonical list.
+# Besides that, there isn't much other use for this.
+# If the curator can directly provide a small canonical list, even a small one will do, this class will not be needed.
+class TagClustering(TagDeduplication):
+    TRAINING_FILE = "curator/clustering_training.json"
+    CLUSTERING_THRESHOLD = 0.1
+
+    def __init__(self):
+        self.deduper = dedupe.Dedupe(TagDeduplication.FIELDS)
+        self.prepare_training()
+
+    # The training data is stored in a file.
+    # If it exists, load from the file
+    # Otherwise, start from scratch
+    def prepare_training(self):
+        data = self.prepare_training_data()
+        if os.path.exists(TagClustering.TRAINING_FILE):
+            with open(TagClustering.TRAINING_FILE, "r") as training_file:
+                self.deduper.prepare_training(data, training_file)
+        else:
+            self.deduper.prepare_training(data)
+
+    # The model is trained and then the data is clustered
+    def cluster_tags(self):
+        self.deduper.train()
+        return self.deduper.partition(
+            self.prepare_training_data(), TagClustering.CLUSTERING_THRESHOLD
+        )
+
+    def save_to_training_file(self):
+        with open(TagClustering.TRAINING_FILE, "w") as file:
+            self.deduper.write_training(file)
+
+    # Saves the clusters to the database
+    def save_clusters(self, clusters):
+        for id_list, confidence_list in clusters:
+            first_tag = Tag.objects.filter(id=id_list[0])
+            if len(first_tag) == 0:
+                continue
+
+            canonical_tag = CanonicalTag.objects.get_or_create(name=first_tag[0].name)
+            for index in range(len(id_list)):
+                tag = Tag.objects.filter(id=id_list[index])
+                confidence = confidence_list[index]
+                if len(tag) > 0:
+                    tag_canon_mapping = CanonicalTagMapping(
+                        tag=tag[0],
+                        canonical_tag=canonical_tag[0],
+                        confidence_score=confidence,
+                    )
+                    tag_canon_mapping.save()
+
+    def training_file_exists(self) -> bool:
+        return os.path.exists(TagClustering.TRAINING_FILE)
+
+    def remove_training_file(self):
+        if os.path.isfile(TagClustering.TRAINING_FILE):
+            os.remove(TagClustering.TRAINING_FILE)
+
+
+class TagGazetteering(TagDeduplication):
+    TRAINING_FILE = "curator/gazetteering_training.json"
+    SEARCH_THRESHOLD = 0.5
```

Comment:
This should probably be parameterizable, with a django settings defined `DEFAULT_TAG_GAZETTEERING_THRESHOLD` that can be overridden when you create an instance, e.g., `TagGazetteering(search_threshold=0.8)`   

```
from django.conf import settings

...

DEFAULT_SEARCH_THRESHOLD = settings.TAG_GAZETTEERING_THRESHOLD
```